Open editors widget in navigator

kenneth-marut-work opened the pull request:
#### What it does

Fixes: #8936 
Adds 'Opened Editors' panel into the Navigator's layout, hidden by default

![grouping-open](https://user-images.githubusercontent.com/62660714/124015206-0797f900-d9aa-11eb-920c-c0137613d522.gif)

Includes the following functionality:
- Shows all opened saveable editors
- Provides SCM, Problem, dirty, and preview state decorators
- Provides caption affix to indicate file's workspace root, as well as relative path
- Adds buttons to close individual editors as well as toolbar buttons to Close All/Save All
- Adds context menu options to tree nodes to perform actions including: Open in Terminal, Copy Path, Save, Compare, Close All, Close Others
- Tracks active editors and will also activate editors when tree nodes are selected, pins preview widget when doubleclicked
- Follows VSCode's grouping convention (i.e. GROUP 1, GROUP 2) to main area editor widgets. If widgets are placed in other shell areas, they are grouped by area (main, bottom, left, right)
- Adds tree node items for "Save All" and "Close All" on group nodes and area nodes
- Will update ordering of widgets when they are rearranged in a tabbar

Known Issues/Areas for Improvement:
- Currently this widget is only tracking NavigatableWidgets (editor widgets) to avoid too much noise, in VSCode it seems that any 'main' area widget is tracked, but this system does not align perfectly with Theia's since we can dock widgets anywhere.
- VSCode provides additional toolbar buttons including 'New Untitled File', and 'Toggle Vertical/Horizontal Layout'
- VSCode will display a badge decoration in the toolbar when there are any unsaved editors:
![image](https://user-images.githubusercontent.com/62660714/113449790-3828c080-93c4-11eb-8c7e-1561868d2eb9.png)
- VSCode auto-sizes its panel layout based on the widget's contents, I'm not quite sure how to achieve this easily with Phosphor's PanelLayout.
- When widgets are rearranged on the left/right toolbars, the ordering will not update (it seems that the left/right DockPanel's `layoutModified` listener behaves differently from the main/bottom dock panel

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Set Up:
- Pull branch and build
- Open a single-root workspace and open some files
- Right click the file explorer's tabbar and select "Open Editors" to enable the widget

Widget Tracking:
- Open some files and observe that editors are tracked and added
- Click on the nodes and observe the their corresponding widgets are activated
- Close widgets from the tabbar, are they removed from the tree view?
- Dock widgets in different panes, split widgets in the main area, observe the grouping strategy. Does this make sense?
- Try out some of the grouping toolbar items like "Close Group" and "Save All in Group"

Decorations:
- Start editing a file, observe the the file is marked as dirty
- If a widget is opened in Preview mode, it should appear in italics (reflecting the Main Area tabbar state) double click on its node to pin the widget, it should be displayed as normal
- Attempt to close the file using the 'x' icon, does it behave as expected?
- Try these steps in a multiroot workspace, observe that caption suffixes are added that indicate the workspace root as well as its relative path
- Introducing invalid syntax in a file should decorate the node using the problem decorator to match its appearance in the file explorer
- Introducing changes will decorate the node with the SCM decorator

Actions:
- Use the toolbar icons and verify they behave as expected
- Use the context menu and verify the items behave as expected


#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,112 @@
+
+    protected updateOpenWidgets = debounce(this.doUpdateOpenWidgets, 250);
+
+    protected async doUpdateOpenWidgets(): Promise<void> {
```
Would there be any value in a more surgical approach here, adding / removing nodes as widgets are opened / closed, rather than rebuilding the tree completely for every widget opening or closing?

kenneth-marut-work replied:
I thought about this as well, but there were a lot of strange timing issues that I came across and it seemed quite difficult to validate removing/adding individual nodes reliably, especially when accounting for the PreviewWidget. In the end I found it easier to rebuild. If you have any suggestions I'm happy to discuss

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,112 @@
+            children: []
+        };
+        for (const widget of openWidgets) {
+            if (Navigatable.is(widget)) {
```
Elsewhere you've checked whether the widget is saveable; why the difference?

kenneth-marut-work replied:
This was from an early attempt at categorizing the widgets. I've changed this to only check for EditorWidget and EditorPreview widget since those are really the only widgets I'm looking for and that ends up being more consistent with what I'm doing elsewhere.

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-widget.tsx:
```diff
@@ -0,0 +1,178 @@
+        this.addClass(OpenEditorsWidget.ID);
+        this.update();
+    }
+    protected activeTreeNodePrefixElement: string | undefined | null;
```
The linter isn't going to like `null`, so you may need a disabling comment.

kenneth-marut-work replied:
Done 👍

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-widget.tsx:
```diff
@@ -0,0 +1,178 @@
+        return React.createElement('div', attributes, content);
+    }
+
+    protected handleMouseEnter = (e: React.MouseEvent<HTMLDivElement>) => this.doHandleMouseEnter(e);
```
Could the logic handled by this enter logic be shifted onto CSS? The way font-awesome and codicon tend to work is by putting some random unicode in a `:after` and then assigning that unicode to the particular icon in their font definition, so it might end up being something like this:

```CSS
.tree-node .left-slot {
    /* normally nothing, so probably no rule */
}

.tree-node.dirty .left-slot:after {
    content: /* unicode corresponding to closed circle in codicon font */
}

.tree-node:hover .left-slot {
   content: /* unicode corresponding to exit icon in codicon font */
}
```

And since you've already added the codicon class, you should get the correct icon.

kenneth-marut-work replied:
Good catch, I knew there must be a better way to do this. I've gone with your second recommendation since I wasn't able to get the unicode approach working

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-file-decorator.ts:
```diff
@@ -0,0 +1,112 @@
+import { EditorPreviewWidget } from '@theia/editor-preview/lib/browser';
+import { Disposable } from '@theia/core/lib/common';
+
+export interface OpenEditorTreeDecorationData extends TreeDecoration.Data {
```
Would it work not to have this as a field on the decoration (which upsets TS a little bit), but evaluate the dirty status in the widget when rendering each node? Then you could fire an event here when the dirty status changed to trigger a rerender in the tree, but not modify the `TreeDecoration.Data` type. Alternatively, you could just move all the dirty-tracking logic to the `TreeModel` or the `Tree` and trigger a rerender in either of those places.

kenneth-marut-work replied:
Good thinking, I've made this change as well. Feels a good bit cleaner, and turns out I was already firing the event for dirty status change

colin-grant-work reviewed (commented):
Looking very nice. I have a few comments on the code

kenneth-marut-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -358,6 +385,26 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+    }
+
+    protected get editorWidgets(): EditorWidget[] {
+        return this.shell.widgets.filter(widget => widget instanceof EditorWidget) as EditorWidget[];
```
Neat trick, I've made that change

kenneth-marut-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -464,6 +518,53 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+
+        // Open Editors Widget Menu Items
+        registry.registerMenuAction(OPEN_EDITORS_CONTEXT_MENU, {
+            commandId: OpenEditorsCommands.CLOSE_ALL_TABS_FROM_TOOLBAR.id,
```
Good catch, that was a leftover from an earlier revision. I've removed it

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-decorator-service.ts:
```diff
@@ -0,0 +1,84 @@
+import { Tree } from '@theia/core/lib/browser';
+
+/**
+ * Symbol for all decorators that would like to contribute into the navigator.
```
Yep, some leftovers from a copy, I've removed both of these comments

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-decorator-service.ts:
```diff
@@ -0,0 +1,84 @@
+        super([...contributions.getContributions(), ...navigatorContributions.getContributions()]);
+    }
+
+    async getDecorations(tree: Tree): Promise<Map<string, TreeDecoration.Data[]>> {
```
Agreed, I may have overcomplicated things a bit. I've modified the logic by being more explicit about what's happening. In reality, we just need to know the colorizing that is coming specifically from the ProblemDecorator, and the apply it to caption suffixes coming from other decorators (in this case, there is only one). Let me know what you think

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-file-decorator.ts:
```diff
@@ -0,0 +1,112 @@
+
+    protected async generateCaptionSuffix(nodeURI: URI): Promise<string> {
+        const workspaceRoots = await this.workspaceService.roots;
+        const parentWorkspace = workspaceRoots.find(({ resource }) => resource.isEqualOrParent(nodeURI));
```
Perfect, I've updated this line to use that utility

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    protected async initializeRoot(): Promise<void> {
+        this.selectionService.onSelectionChanged(selection => {
```
Good catch, fixed

colin-grant-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -358,6 +385,26 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
         });
+        registry.registerCommand(OpenEditorsCommands.CLOSE_ALL_TABS_FROM_TOOLBAR, {
+            execute: widget => this.withOpenEditorsWidget(widget, async () => {
+                this.shell.widgets.forEach(w => {
```
I know we've talked about this a bit, but I think perhaps the `.all` field on the `EditorManager` would give you these? It's supposed to track all the widgets created by the `EditorManager` that are still open, which should include those that are open inside an `EditorPreviewWidget`. But you may have already told me why that doesn't work...

kenneth-marut-work replied:
I had originally used that but ended up favoring the shell approach since It's possible in the future we might want to include widgets other than EditorWidgets in the tree. Ultimately I just decided to remove the dependency on the Editor package and using `NavigatableWidget.is` instead.

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,112 @@
+        this.fireChanged();
+    }
+
+    protected updateOpenWidgets = debounce(this.doUpdateOpenWidgets, 250);
```
I believe trailing: true is applied as the default

vince-fugnitto commented on packages/navigator/package.json:
```diff
@@ -4,6 +4,7 @@
   "description": "Theia - Navigator Extension",
   "dependencies": {
     "@theia/core": "1.12.0",
+    "@theia/editor-preview": "1.12.0",
```
Ideally we should not introduce new coupling between components, especially since nothing depends on `@theia/editor-preview` so applications currently have the flexibility of including it or not.

kenneth-marut-work replied:
I was able to remove this coupling in my latest commit 👍

vince-fugnitto commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -158,6 +163,28 @@ export namespace NavigatorContextMenu {
+        iconClass: 'codicon codicon-save-all'
+    };
+
+    export const SAVE_EDITOR: Command = {
+        id: 'navigator.save.editor',
+        category: 'File',
+        label: 'Save'
+    };
```
I believe the command is unused.

kenneth-marut-work replied:
Good catch, I've removed this command.

vince-fugnitto reviewed (commented):
I did a high level pass of the pull-request, the main concern was the new coupling introduced by the changes. 

I found an issue with the current styling, for different icon-themes (such as the seti) we encounter some problems:


https://user-images.githubusercontent.com/40359487/115753859-a4666680-a369-11eb-88d3-70e90147215d.mp4

kenneth-marut-work commented:
> I did a high level pass of the pull-request, the main concern was the new coupling introduced by the changes.
> 
> I found an issue with the current styling, for different icon-themes (such as the seti) we encounter some problems:
> 
>  open-editors-ui.mp4

![icon-themes](https://user-images.githubusercontent.com/62660714/115912387-dee90580-a434-11eb-864f-e128c86e52a7.gif)

I've updated the styling to satisfy changing the file icon themes. There is a bit of odd behavior when switching to Seti in particular as the indentation is not consistent with other themes (you can see this on the navigator tree as well), but does not cause any overlap with the decorator. Let me know if this is sufficient. Thanks!

kenneth-marut-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -158,6 +163,28 @@ export namespace NavigatorContextMenu {
+    export const SAVE_ALL_TABS: Command = {
+        id: 'navigator.save.all.editors',
+        category: 'File',
+        label: 'Save All Editor',
```
I've fixed the typo and also kept the label in place since I am using it for the toolbar icon's tooltip. Since the command handler has `withOpenEditorsWidget` check, it should not be showing up in the global palette.

colin-grant-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -358,6 +382,20 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+        });
+    }
+
+    protected get editorWidgets(): NavigatableWidget[] {
```
Rather than putting this here, could you have a field on the `OpenEditorsWidget` itself that tracks the number of widgets it contains and access that, instead? It would save a couple of filter calls, at least. It should be `get`able as something like `this.model.root?.children?.length ?? 0`, though I'm sure it'll want a `CompositeTreeNode.is()` check rather than just pretending that `children` might exist on *any* old tree node!

kenneth-marut-work replied:
I've kept the getter here for convenience but exposed a getter through the tree model that retrieves the editorWidgets. I also had to use the `widgetManager` (rather than `this.widget`) to access the `OpenEditorsWidget` since the `navigator-contribution`'s widget is actually the `FileNavigatorWidget`

colin-grant-work reviewed (commented):
This seems to be working well for me, and the dependencies on the `editor` and `editor-preview` packages have been removed. One small hiccup is the size of widget - it doesn't seem to be adhering to the weight ratio specified - and the handling of multiple instances of the same editor widget.

kenneth-marut-work commented:
> This seems to be working well for me, and the dependencies on the `editor` and `editor-preview` packages have been removed. One small hiccup is the size of widget - it doesn't seem to be adhering to the weight ratio specified - and the handling of multiple instances of the same editor widget.

Thanks for taking another look @colin-grant-work. I've addressed your feedback, and also left a comment in the code regarding the quoted issue above.

kenneth-marut-work commented:
@colin-grant-work  I've pushed an update to this PR that adds grouping functionality to the OpenEditors widget based on my best attempt at reconciling VSCode's OpenEditorsWidget behavior with Theia's ability to dock widgets anywhere. I've also added the corresponding toolbar icons to the grouping nodes that allow for "Close Group" and "Save all in group", and added listeners so that the tree is reordered when tabs are rearranged. Please give this a try and let me know if you have any suggestions

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    protected setupHandlers(): void {
+        this.toDispose.push(this.selectionService.onSelectionChanged(selection => {
+            const { widget } = (selection[0] as OpenEditorNode);
+            this.applicationShell.activateWidget(widget.id);
+        }));
```
It looks like this is basically intended to be a click handler `handleClickEvent`? If so, one option would be to override the click handler in the tree widget, as is done in the `navigator-widget` code.

Testing this also made me notice that it doesn't look like selection is reflected in the tree. I only see hover marking, not selected / active marking.

kenneth-marut-work replied:
Good call, I've used the tree widget's selection method to do the activation. Even though the regular NavigatorTree uses "revealWidget" to activate the widget, I've gone activateWidget, since there's no obvious visual indication of which widget is being selected, especially when multiple groups are open. Also, I'm relying on the shell's knowledge of `activeWidget` to execute some of context menu commands reliably

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+            const { widget } = (selection[0] as OpenEditorNode);
+            this.applicationShell.activateWidget(widget.id);
+        }));
+        this.toDispose.push(this.applicationShell.onDidChangeCurrentWidget(async ({ newValue }) => {
+            const nodeToSelect = this.tree.getNode(newValue?.id) as SelectableTreeNode;
+            if (nodeToSelect) {
+                this.selectNode(nodeToSelect);
+            }
+        }));
```
See above re: selection not appearing / not persisting. It looks like the correct node is selected, but the selection color disappears after a brief flash.

kenneth-marut-work replied:
This one was tricky, the `layoutModified` event I had been listening to to check for tabbar reordering also fires when the currentTitle is changed, this seemed to be causing the flicker. 

There didn't seem to be a better (less complicated) way of checking for reordering, so I've added a check that runs during `layoutModified` events, which manually compares the lastOrdering with the currentOrdering. This seems to be reliable

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+        }));
+        this.toDispose.push(this.workspaceService.onWorkspaceChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.workspaceService.onWorkspaceLocationChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.applicationShell.onDidAddWidget(widget => {
+            if (ApplicationShell.TrackableWidgetProvider.is(widget) && widget.onDidChangeTrackableWidgets) {
+                this.toDisposeOnPreviewWidgetReplaced.dispose();
+                this.toDisposeOnPreviewWidgetReplaced.push(widget.onDidChangeTrackableWidgets(() => this.updateOpenWidgets()));
+            }
+            this.updateOpenWidgets();
+        }));
```
Looks like some of this refers to the old `EditorPreview` implementation.

kenneth-marut-work replied:
Thanks, I was able to remove this entirely 👌

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+                this.selectNode(nodeToSelect);
+            }
+        }));
+        this.toDispose.push(this.workspaceService.onWorkspaceChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.workspaceService.onWorkspaceLocationChanged(() => this.updateOpenWidgets()));
```
I see how these events change the decorations, but I'm not sure how they would affect widgets that are open?

kenneth-marut-work replied:
Good catch, I've removed these

colin-grant-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+
+    protected async doUpdateOpenWidgets(): Promise<void> {
+        this._editorWidgetsByArea = new Map<ApplicationShell.Area, NavigatableWidget[]>();
+        const areas: ApplicationShell.Area[] = ['main', 'bottom', 'left', 'right'];
```
I'm sure you were tempted to change this to an `enum` rather than a string union type. Maybe it should be done? I also notice you've omitted `top`. Is that intentional?

kenneth-marut-work replied:
I had thought about that, but there are enough placed where this type is used to make me a bit nervous to do that now. I did intentionally leave 'top' out because I believe that area refers to the browser versions top tabbar (File menu) and didn't expect to find any widgets there. But I've added it back in since it shouldn't really make any difference.

colin-grant-work commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -408,6 +409,10 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
             commandId: TerminalCommands.TERMINAL_CONTEXT.id,
             order: 'z'
         });
+        menus.registerMenuAction(TerminalMenus.TERMINAL_OPEN_EDITORS_CONTEXT_MENU, {
+            commandId: TerminalCommands.TERMINAL_CONTEXT.id,
+            order: 'z'
+        });
```
This command is news to me. :-)

kenneth-marut-work replied:
Yes, same here. Pretty handy. It opens the terminal to the nearest directory path of the file.

colin-grant-work reviewed (commented):
Looking pretty good. The behavior is pretty much spot on, just some minor comments about small display items and some stuff in the code.

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-decorator-service.ts:
```diff
@@ -0,0 +1,80 @@
+export const OpenEditorsTreeDecorator = Symbol('OpenEditorsTreeDecorator');
+
+@injectable()
+export class OpenEditorsTreeDecoratorService extends AbstractTreeDecoratorService {
```
Looks like there's already an issue about that here #1564. I'll probably skip on implementing this feature in this PR. Good catch though

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-widget.tsx:
```diff
@@ -0,0 +1,207 @@
+            {this.renderIndent(node, { depth })}
+            {this.renderNode(node, { depth })}
+        </div>;
+    }
```
I've removed the render indent call 👉

kenneth-marut-work commented on packages/editor-preview/src/browser/editor-preview-tree-decorator.ts:
```diff
@@ -0,0 +1,127 @@
+    protected readonly toDisposeOnDirtyChanged = new Map<string, Disposable>();
+
+    onDidInitializeLayout(app: FrontendApplication): void {
+        this.shell = app.shell;
```
Good call, some of this had been copied from the problem decorator I believe, but it seems unnecessary. I've switched it to the ordinary injection of ApplicationShell

kenneth-marut-work commented on packages/navigator/src/browser/navigator-frontend-module.ts:
```diff
@@ -56,6 +61,15 @@ export default new ContainerModule(bind => {
+    bindContributionProvider(bind, NavigatorTreeDecorator);
+    bindContributionProvider(bind, OpenEditorsTreeDecorator);
+
+    bind(OpenEditorsWidget).toSelf().inSingletonScope();
```
Good catch, removed this binding

kenneth-marut-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -357,6 +367,38 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+                    saveable?.save();
+                });
+            }
+        });
```
I've set the visibility of these commands to false. It seems like they are visible in VSCode, but it doesn't seem like it makes much sense since there's no reference to the Group from the command palette.

colin-grant-work commented on packages/editor-preview/src/browser/editor-preview-tree-decorator.ts:
```diff
@@ -40,37 +40,48 @@ export class EditorPreviewTreeDecorator implements TreeDecorator, FrontendApplic
         const saveable = Saveable.get(widget);
-        const isTrackableProvider = ApplicationShell.TrackableWidgetProvider.is(widget);
-        if (saveable && !isTrackableProvider) {
+        if (saveable && widget instanceof EditorPreviewWidget) {
```
You may want to have two blocks for this, one for saveable and one for editor preview, since the two aren't guaranteed to coincide.

kenneth-marut-work replied:
great suggestion, fixed

colin-grant-work commented on packages/editor-preview/src/browser/editor-preview-tree-decorator.ts:
```diff
@@ -40,37 +40,48 @@ export class EditorPreviewTreeDecorator implements TreeDecorator, FrontendApplic
+    protected unregisterEditorListeners(widget: Widget): void {
+        this.toDisposeOnDirtyChanged.get(widget.id)?.dispose();
+        this.toDisposeOnDirtyChanged.delete(widget.id);
+        if (widget instanceof EditorPreviewWidget && widget.isPreview) {
+            this.toDisposeOnPreviewPinned.get(widget.id)?.dispose();
+            this.toDisposeOnDirtyChanged.delete(widget.id);
```
Mixed references here to `toDisposeOnPreviewPinned` and `toDisposeOnDirtyChanged`.

It looks like the logic above should ensure that you'll only have an entry in `toDisposeOnPreviewPinned` if the editor still `.isPreview`, but since there's no harm in trying to `.get` its `.id` or `.delete` it if it doesn't exist, I might remove the conditional anyway.

kenneth-marut-work replied:
Thanks for catching that, I've done both of these things

colin-grant-work reviewed (approved):
This is working well to me, and the code looks reasonable. My greatest desideratum would be to follow VSCode in resizing the tree to only ever take up the amount of space it needs, but that can be tackled in a follow up PR.

kenneth-marut-work commented:
Hey @vince-fugnitto, I've made some considerable updates to this MR, would you be able to give it a test whenever you have free time?  

>My greatest desideratum would be to follow VSCode in resizing the tree to only ever take up the amount of space it needs, but that can be tackled in a follow up PR.

Regarding @colin-grant-work's comment above about the automatic resizing of a ViewContainer part, do you (@vince-fugnitto) have any suggestions on how to do that with Phosphor's built in methods? I couldn't seem to find a convenient way to do this.

Thanks!

vince-fugnitto commented:
@kenneth-marut-work do you mind rebasing, and I'd be happy to take a look :+1: As for the size of the view, it doesn't need to be part of this pull-request, I believe all tree-view parts suffer from the same bug where the size is not controllable.

vince-fugnitto commented on packages/editor-preview/package.json:
```diff
@@ -4,7 +4,9 @@
   "dependencies": {
     "@theia/core": "1.15.0",
-    "@theia/editor": "1.15.0"
+    "@theia/editor": "1.15.0",
+    "@theia/navigator": "1.15.0",
+    "@theia/workspace": "1.15.0"
```
@kenneth-marut-work @colin-grant-work is the new coupling necessary?
`@theia/editor` only uses `core`, why does `editor-preview` need the coupling?

colin-grant-work replied:
@vince-fugnitto, I believe that these were added because `editor-preview` now contributes decorations to the open-editors widget that determine whether the workspace is shown as part of the node suffixes and whether the caption should be italicized (preview) or not (not preview).

kenneth-marut-work replied:
The two of us discussed this since the tree decorator (which is responsible for italicizing the filename when in preview mode, and also adding workspace/tail decorations) was decided to be moved from the navigator -> editor preview package. This allows the decorator to have access to the editor-preview's built-in checks for preview state, and emitters for on-preview-pinned. 

It has the sideeffect of needing to depend on the navigator package (to access the decorator service) as well as the workspace service. The SCM and markers package have a similar dependency set up.

vince-fugnitto commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-commands.ts:
```diff
@@ -0,0 +1,47 @@
+    };
+
+    export const SAVE_ALL_IN_GROUP: Command = {
+        id: 'navigator.save.all.in.ground',
```
typo?

```suggestion
        id: 'navigator.save.all.in.group',
```

kenneth-marut-work replied:
Thanks, I've fixed this

vince-fugnitto commented on packages/navigator/src/browser/open-editors-widget/open-editors.css:
```diff
@@ -0,0 +1,84 @@
+    font-size: var(--theia-ui-font-size0);
+}
+
+.open-editors-node-row .open-editors-prefix-icon-container {
```
I believe we should prefix the new stylings with `theia-open-editors-widget`

kenneth-marut-work replied:
👌

vince-fugnitto commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -357,6 +367,40 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+                const widgets = openEditorsWidget.getEditorWidgetsByGroup(id);
+                widgets?.forEach(widget => widget.close());
+            },
+            isVisible: () => false
```
To confirm, the following is set to false to hide it from the `command palette`?

kenneth-marut-work replied:
So these two commands I've decided to hide from the command palette since they are essentially toolbar icon commands. Their implementation is slightly different from the way VSCode would implement them as they accept either an `ApplicationShell.Area` or `TabBar<Widget>` command as their argument.

vince-fugnitto replied:
I thought a simpler way would be to not give it a `label` so it does not show up in the command palette?
Although, I'm fine with you explicitly setting it to `false` as well.

vince-fugnitto commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-commands.ts:
```diff
@@ -0,0 +1,47 @@
+
+export namespace OpenEditorsCommands {
+    export const CLOSE_ALL_TABS_FROM_TOOLBAR: Command = {
+        id: 'navigator.close.all.editors',
```
For the new commands in the file, should we align their `ids` with that of vscode so the plugin-system can directly reference them without the mapping?

kenneth-marut-work replied:
See note above, I've kept these commands separate from the plugin commands since their implementation is slightly different in both arguments and behavior. I've added added an additional command with VSCode's ID to the `plugin-vscode-commands-contribution` for `workbench.files.saveAllInGroup` which uses a very similar implementation as the existing `workbench.action.closeEditorsInGroup` and ensured that the two now have `labels` so that they can be accessed from the command palette. It's a bit tricky to figure out exactly where some of these things should live, but this was my best attempt at separating responsibilities. I've also added an appropriate utility in the application shell for saving editors which I am using in the `execute` of the toolbar commands.

Let me know if you have suggestions, thanks!

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+    get editorWidgets(): NavigatableWidget[] {
+        let editorWidgets: NavigatableWidget[] = [];
+        this._editorWidgetsByArea.forEach(widgets => {
+            editorWidgets = [...editorWidgets, ...widgets];
```
👍

kenneth-marut-work commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts:
```diff
@@ -0,0 +1,218 @@
+        return this._editorWidgetsByArea.get(id as ApplicationShell.Area);
+    }
+
+    protected _editorWidgetsByGroup = new Map<number, NavigatableWidget[]>();
```
👍

vince-fugnitto commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -105,7 +113,8 @@ export namespace FileNavigatorCommands {
     export const COPY_RELATIVE_FILE_PATH: Command = {
-        id: 'navigator.copyRelativeFilePath'
+        id: 'navigator.copyRelativeFilePath',
+        label: 'Copy Relative Path'
```
Should we align?

```suggestion
        label: 'Copy Relative Path of Active File'
```

kenneth-marut-work replied:
In VSCode, when I right click on files and look for this command I see "Copy Relative Path". I do see two labels for this command in the VSCode repo, but I'm not sure I understand how VSCode represents their commands to make the call:

https://github.com/microsoft/vscode/blob/2853fd0ffcc4c067bf7b5b31ccfc15ac245fc89c/src/vs/workbench/contrib/files/browser/fileActions.contribution.ts#L163

https://github.com/microsoft/vscode/blob/2853fd0ffcc4c067bf7b5b31ccfc15ac245fc89c/src/vs/workbench/contrib/files/browser/fileActions.contribution.ts#L215

vince-fugnitto commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -357,6 +369,38 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
+            isVisible: widget => this.withOpenEditorsWidget(widget, () => !!this.editorWidgets.length)
+        });
+        registry.registerCommand(OpenEditorsCommands.SAVE_ALL_TABS_FROM_TOOLBAR, {
+            execute: widget => this.withOpenEditorsWidget(widget, () => this.commandService.executeCommand(CommonCommands.SAVE_ALL.id)),
```
No need for the `CommandService` injection:

```suggestion
            execute: widget => this.withOpenEditorsWidget(widget, () => registry.executeCommand(CommonCommands.SAVE_ALL.id)),
```

kenneth-marut-work replied:
Good to know, I've fixed this 👍

vince-fugnitto commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -412,7 +463,7 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
         registry.registerMenuAction(NavigatorContextMenu.CLIPBOARD, {
             commandId: FileNavigatorCommands.COPY_RELATIVE_FILE_PATH.id,
-            label: 'Copy Relative Path',
+            label: FileNavigatorCommands.COPY_RELATIVE_FILE_PATH.label,
```
I believe it can be omitted if the `label` is the same as the `command` definition.
The same is true for other occurences.

kenneth-marut-work replied:
Also good to know, I've removed the labels from the menu items

vince-fugnitto commented on packages/navigator/src/browser/open-editors-widget/navigator-open-editors-decorator-service.ts:
```diff
@@ -0,0 +1,80 @@
+export const OpenEditorsTreeDecorator = Symbol('OpenEditorsTreeDecorator');
+
+@injectable()
+export class OpenEditorsTreeDecoratorService extends AbstractTreeDecoratorService {
```
I'm a bit confused as to this file. 
1. why does it live in `navigator`, `navigator` should not know about `problems`.
2. I believe we want a `open-editor` decorator service which `problems` contributes to, like the following in `problems/`.
```ts
bind(NavigatorTreeDecorator).toService(ProblemDecorator);
```

kenneth-marut-work replied:
I've moved the suffix decorations that add the `workspace * path/to/file` to the problem decorator and added a conditional check if the decorated node belongs to the OpenEditorsWidget. This way the suffix decorations are only applied to open-editor nodes and not the navigator nodes. Moving that logic allows this file to be much less complex :)

Thanks for the suggestions and help with this issue!

The only downside is that I cannot find a way to aggregate the italicization provided by the `editor-preview-tree-decorator` into the suffix decoration, which causes a preview widget node to not have italicization on its suffix. I don't think this is really a problem and the cleaner code outweighs the minor UI flaw. This could also be addressed in a separate PR

![image](https://user-images.githubusercontent.com/62660714/126011162-cffafe00-80dd-4e64-ab03-b18c71fc7209.png)

kenneth-marut-work commented:
@vince-fugnitto thanks for the helpful suggestions. I believe I have addressed your comments except for the minor issue with the `visible: () => false` check which I've decided to keep as I refer to those commands' labels by their Variable name in the OpenEditorsWidget for convenience.

Moving the suffix-decorations into the problem-decorator seemed to work quite well and greatly simplified the strange logic I was doing in the open-editors-decorator service to retroactively apply coloring, so now that file is much more straightforward.

kenneth-marut-work commented:
@vince-fugnitto thanks for pointing out the path display issue. I've fixed this and also added a case for when a file is opened outside a workspace (in that case we display the full path). This works for the case of no-workspace-open case as well:

**Single root with nested files and file outside workspace**
![image](https://user-images.githubusercontent.com/62660714/126347828-7afc5172-ef21-4103-808b-0616512eee4c.png)

**Multiroot with nested files, and file outside workspace**
![image](https://user-images.githubusercontent.com/62660714/126347252-5cd7a1fb-a0d7-4bdd-aa00-5f1d80210b60.png)

**No-root workspace with file**
![image](https://user-images.githubusercontent.com/62660714/126347454-4110f965-95cb-427e-8108-08ed08fd27d2.png)

vince-fugnitto reviewed (approved):
The changes work well for me, and the code is also clean 👍 

I attempted to review multiple use-cases:
- [x] confirmed it works in a single-root workspace
- [x] confirmed it works in a multi-root workspace
- [x] confirmed the relative paths are displayed for nested files
- [x] confirmed the root name and relative paths are displayed for files in a multi-root workspace
- [x] confirmed the problem decorations work for the open editors
- [x] confirmed that scm decorations work for the open editors
- [x] confirmed that dirty decorations work for the open editors
- [x] confirmed that the close-all and save-all toolbar items work
- [x] confirmed that the context-menu works for open editors
- [x] confirmed that switching editors in the view works
- [x] confirmed that it properly represents editor groups in the layout

The changes are already substantial, if there are improvements we can make them in subsequent pull-requests 👍

kenneth-marut-work commented:
> The changes work well for me, and the code is also clean 👍
> 
> I attempted to review multiple use-cases:
> 
> * [x]  confirmed it works in a single-root workspace
> * [x]  confirmed it works in a multi-root workspace
> * [x]  confirmed the relative paths are displayed for nested files
> * [x]  confirmed the root name and relative paths are displayed for files in a multi-root workspace
> * [x]  confirmed the problem decorations work for the open editors
> * [x]  confirmed that scm decorations work for the open editors
> * [x]  confirmed that dirty decorations work for the open editors
> * [x]  confirmed that the close-all and save-all toolbar items work
> * [x]  confirmed that the context-menu works for open editors
> * [x]  confirmed that switching editors in the view works
> * [x]  confirmed that it properly represents editor groups in the layout
> 
> The changes are already substantial, if there are improvements we can make them in subsequent pull-requests 👍

Thanks @vince-fugnitto, I'll go ahead and squash and otherwise will merge tomorrow if there are no objections
